crates/ark/src/lsp/completions/sources/composite.rs

```diff
@@ -14,25 +14,33 @@ mod snippets;
 mod subset;
 mod workspace;
 
-use std::collections::HashSet;
+use std::collections::HashMap;
 
 use stdext::*;
 use tower_lsp::lsp_types::CompletionItem;
 use tower_lsp::lsp_types::CompletionItemKind;
 use tree_sitter::Node;
 
 use crate::lsp::completions::completion_context::CompletionContext;
-use crate::lsp::completions::sources::push_completions;
+use crate::lsp::completions::sources::collect_completions;
+use crate::lsp::completions::sources::CompletionSource;
 use crate::treesitter::NodeType;
 use crate::treesitter::NodeTypeExt;
 
+// Locally useful data structure for tracking completions and their source
+#[derive(Clone, Default)]
+struct CompletionItemWithSource {
+    item: CompletionItem,
+    source: String,
+}
+
 /// Gets completions from all composite sources, with deduplication and sorting
 pub(crate) fn get_completions(
     completion_context: &CompletionContext,
 ) -> anyhow::Result<Option<Vec<CompletionItem>>> {
     log::info!("Getting completions from composite sources");
 
-    let mut completions: Vec<CompletionItem> = vec![];
+    let mut completions: HashMap<String, CompletionItemWithSource> = HashMap::new();
```

Comment:
```suggestion
    let mut completions = HashMap::new();
```

I bet the compiler "just knows"